include/pybind11/common.h

```diff
@@ -321,6 +321,7 @@ struct internals {
     std::unordered_map<const void *, void*> registered_types_py;       // PyTypeObject* -> type_info
     std::unordered_multimap<const void *, void*> registered_instances; // void * -> PyObject*
     std::unordered_set<std::pair<const PyObject *, const char *>, overload_hash> inactive_overload_cache;
+    std::unordered_map<std::type_index, std::vector<bool (*)(PyObject *, void *&)>> direct_conversions;
```

Comment:
The only potential issue I see is that ``register_dtype`` is called before the associated ``class_<>`` constructor has executed. Like with inheritance and similar features that depend on some prior initialization, I think that it's fair to impose a mandatory ordering here to keep the library internals simple.